### PR TITLE
"function factory" in r_draw.c and v_video.c

### PR DESCRIPTION
**Pull request by rfomin**
Alternative to #1610

### CODE REVIEW

File: src/r_draw.c [hunk 1]
```diff
@@ -92,97 +95,76 @@ byte    dc_skycolor;
...
-          byte src = source[(frac>>FRACBITS) & heightmask];
-          *dest = colormap[brightmap[src]][src];
-        }
-      }
-  }
-} 
+R_DRAW(Column, colormap[0][src])
+R_DRAW(ColumnDark, darkcolormap[colormap[0][src]])
+R_DRAW(ColumnBrightmap, colormap[dc_brightmap[src]][src])
+R_DRAW(ColumnBrightmapDark, darkcolormap[colormap[dc_brightmap[src]][src]])
```
Comment (rfomin) [inline src/r_draw.c:167]: Such large macro "functions" are bad style, but they reduce code duplication.

**Comment by rfomin** [comment]
This is an unfinished, I'm interested in your opinion, should I continue? Maybe we should just copy and paste these functions, or maybe refactor the non performance-critical parts.

**Comment by ceski-1** [comment]
> This is unfinished, I'm interested in your opinion, should I continue? Maybe we should just copy and paste these functions, or refactor the non performance-critical parts.

The alternative could be something like this right? But possibly a performance impact?

<details><summary>pseudocode</summary>

```c
static byte *dc_dest;
typedef byte (*map_func_t)(byte src);

static byte MapTLColumnBrightmapDark(byte src)
{
  return darkcolormap[tranmap[(*dc_dest << 8) + dc_colormap[dc_brightmap[src]][src]]];
}

void DrawColumn(map_func_t map_func)
{
  dc_dest = ylookup[dc_yl] + columnofs[dc_x];
  src = source[frac >> FRACBITS];
  *dc_dest = map_func(src);
}

static void DrawTLColumnBrightmapDark(void)
{
  DrawColumn(&MapTLColumnBrightmapDark);
}

void (*R_DrawTLColumn)(void);

void R_InitDrawFunctions(void)
{
  R_DrawTLColumn = DrawTLColumnBrightmapDark;
}
```

</summary>

**Comment by rfomin** [comment]
> The alternative could be something like this right? But possibly a performance impact?

Yes, of course, then there will be a lot of function calls. Compiler will probably inline most of them. Perhaps we should test this.

**Comment by rfomin** [comment]
I've done some reading about the `inline` keyword, it seems that function pointer calls cannot be inlined, the compiler has to create object code for the function pointer. In general `inline` is finicky and compiler dependent, I think it's a bad idea to rely on it.

I did a little experiment with pointers to inline functions and got a significant performance hit with MSVC and MSYS Clang (about 30%), so it didn't work.

The main disadvantage of the macros method in this PR is problems with the use of debugger. I'm using [clangd](https://clangd.llvm.org/) with a very nice "expand macro" feature, so this problem can be avoided. @fabiangreffrath @ceski-1 What do you think? I'd like to continue with the macros approach.

Function pointers patch:
<details>

```diff
diff --git a/src/r_draw.c b/src/r_draw.c
index 727c9760..1b2ed954 100644
--- a/src/r_draw.c
+++ b/src/r_draw.c
@@ -153,10 +153,102 @@ byte    dc_skycolor;
         }                                                              \
     }
 
-DRAW_COLUMN(, dc_colormap[0][src])
-DRAW_COLUMN(Dark, darkcolormap[dc_colormap[0][src]])
-DRAW_COLUMN(Brightmap, dc_colormap[dc_brightmap[src]][src])
-DRAW_COLUMN(BrightmapDark, darkcolormap[dc_colormap[dc_brightmap[src]][src]])
+#if defined(_MSC_VER)
+#  define INLINE __forceinline
+#else
+#  define INLINE inline
+#endif
+
+static INLINE byte GetPixel(byte src)
+{
+    return dc_colormap[0][src];
+}
+
+static INLINE byte GetPixelDark(byte src)
+{
+    return darkcolormap[dc_colormap[0][src]];
+}
+
+static INLINE byte GetPixelBrightmap(byte src)
+{
+    return dc_colormap[dc_brightmap[src]][src];
+}
+
+static INLINE byte GetPixelBrightmapDark(byte src)
+{
+    return darkcolormap[dc_colormap[dc_brightmap[src]][src]];
+}
+
+static byte (*srcpixel)(byte) = GetPixel;
+
+static void DrawColumn(void)
+{
+    int count = dc_yh - dc_yl + 1;
+
+    if (count <= 0)
+    {
+        return;
+    }
+
+    byte *dest = ylookup[dc_yl] + columnofs[dc_x];
+    const fixed_t fracstep = dc_iscale;
+
+    fixed_t frac = dc_texturemid + (dc_yl - centery) * fracstep;
+    int heightmask = dc_texheight - 1;
+
+    if (dc_texheight & heightmask)
+    {
+        heightmask++;
+        heightmask <<= 16;
+
+        if (frac < 0)
+        {
+            while ((frac += heightmask) < 0)
+                ;
+        }
+        else
+        {
+            while (frac >= heightmask)
+            {
+                frac -= heightmask;
+            }
+        }
+
+        do
+        {
+            byte src = dc_source[frac >> 16];
+            *dest = srcpixel(src);
+            dest += linesize;
+            if ((frac += fracstep) >= heightmask)
+            {
+                frac -= heightmask;
+            }
+        } while (--count);
+    }
+    else
+    {
+        while ((count -= 2) >= 0)
+        {
+            byte src = dc_source[(frac >> 16) & heightmask];
+            *dest = srcpixel(src);
+            dest += linesize;
+            frac += fracstep;
+            src = dc_source[(frac >> 16) & heightmask];
+            *dest = srcpixel(src);
+            dest += linesize;
+            frac += fracstep;
+        }
+        if (count & 1)
+        {
+            byte src = dc_source[(frac >> 16) & heightmask];
+            *dest = srcpixel(src);
+        }
+    }
+}
+
+//DRAW_COLUMN(Dark, darkcolormap[dc_colormap[0][src]])
+//DRAW_COLUMN(Brightmap, dc_colormap[dc_brightmap[src]][src])
+//DRAW_COLUMN(BrightmapDark, darkcolormap[dc_colormap[dc_brightmap[src]][src]])
 
 // Here is the version of R_DrawColumn that deals with translucent  // phares
 // textures and sprites. It's identical to R_DrawColumn except      //    |
@@ -652,7 +744,31 @@ R_DRAW_SPAN(Dark, darkcolormap[ds_colormap[0][src]])
 R_DRAW_SPAN(Brightmap, ds_colormap[ds_brightmap[src]][src])
 R_DRAW_SPAN(BrightmapDark, darkcolormap[ds_colormap[ds_brightmap[src]][src]])
 
-void (*R_DrawColumn)(void) = DrawColumn;
+static void DrawColumnAlt(void)
+{
+    srcpixel = GetPixel;
+    DrawColumn();
+}
+
+static void DrawColumnAltDark(void)
+{
+    srcpixel = GetPixelDark;
+    DrawColumn();
+}
+
+static void DrawColumnAltBrightmap(void)
+{
+    srcpixel = GetPixelBrightmap;
+    DrawColumn();
+}
+
+static void DrawColumnAltBrightmapDark(void)
+{
+    srcpixel = GetPixelBrightmapDark;
+    DrawColumn();
+}
+
+void (*R_DrawColumn)(void) = DrawColumnAlt;
 void (*R_DrawTLColumn)(void) = DrawColumnTL;
 void (*R_DrawTranslatedColumn)(void) = DrawColumnTR;
 void (*R_DrawSpan)(void) = DrawSpan;
@@ -667,28 +783,28 @@ void R_InitDrawFunctions(void)
 
     if (local_brightmaps && shade_screen)
     {
-        R_DrawColumn = DrawColumnBrightmapDark;
+        R_DrawColumn = DrawColumnAltBrightmapDark;
         R_DrawTLColumn = DrawColumnTLBrightmapDark;
         R_DrawTranslatedColumn = DrawColumnTRBrightmapDark;
         R_DrawSpan = DrawSpanBrightmapDark;
     }
     else if (local_brightmaps && !shade_screen)
     {
-        R_DrawColumn = DrawColumnBrightmap;
+        R_DrawColumn = DrawColumnAltBrightmap;
         R_DrawTLColumn = DrawColumnTLBrightmap;
         R_DrawTranslatedColumn = DrawColumnTRBrightmap;
         R_DrawSpan = DrawSpanBrightmap;
     }
     else if (shade_screen && !local_brightmaps)
     {
-        R_DrawColumn = DrawColumnDark;
+        R_DrawColumn = DrawColumnAltDark;
         R_DrawTLColumn = DrawColumnTLDark;
         R_DrawTranslatedColumn = DrawColumnTRDark;
         R_DrawSpan = DrawSpanDark;
     }
     else
     {
-        R_DrawColumn = DrawColumn;
+        R_DrawColumn = DrawColumnAlt;
         R_DrawTLColumn = DrawColumnTL;
         R_DrawTranslatedColumn = DrawColumnTR;
         R_DrawSpan = DrawSpan;

```
</details>

**Comment by ceski-1** [comment]
> The main disadvantage of the macros method in this PR is problems with the use of debugger. I'm using [clangd](https://clangd.llvm.org/) with a very nice "expand macro" feature, so this problem can be avoided.

I've been relying on something similar in vscode which is helpful. In general, I struggle with macros this large but I don't have a strong opinion either way.

**Comment by rfomin** [comment]
> In general, I struggle with macros this large but I don't have a strong opinion either way.

I don't like macros in C, but I think we can make an exception for these functions. A huge file with mostly copied and pasted code is also not very readable.

**Comment by fabiangreffrath** [comment]
One advantage I see in these macros is that they are easier to adapt to (hypothetic, in the distant future) truecolor rendering.

**Comment by ceski-1** [comment]
Agreed, in this case it may be worth the trade-off.

**Comment by fabiangreffrath** [comment]
I like the way you cleaned up the drawing functions! But I'm really not sure if fixing #1609 is worth introducing the `R_ShadeScreen()` and `V_ShadeScreen()` indirections.

**Comment by rfomin** [comment]
> But I'm really not sure if fixing #1609 is worth introducing the `R_ShadeScreen()` and `V_ShadeScreen()` indirections.

I thought why not, if we have "function generation" anyway. This not only solves the HOM issue, but also improves performance, because now we do everything in one go and we don't have to fill the screen twice with pixels for shading.

**Comment by ceski-1** [comment]
At native resolution on a 2560x1440 display, on map01 with the overlap map active, there's a nice increase in frame rate from 133 fps with the master branch to 156 fps with this branch.

I tested a wad with a bunch of random effects: [effects.zip](https://github.com/user-attachments/files/15604430/effects.zip)
Everything is fine except fuzz, but I think some other recent commit did something to it.

**Pull request merged**